Added specMACS cloud geometry markdown example and pictures

lvol08 opened the pull request.

d70-t commented:
Hey @lvol08, thanks for this valuable contribution 🎉 

I noticed, that your markdown doesn't seem to be a [MySt](https://myst-parser.readthedocs.io/en/latest/) notebook. Furthermore, it contains code output and pre-generated plots. For the HowTo Eurec4a, we want to run all the code described by the book using CI pipelines, in order to verify that the code actually does what it should do. [Here's](https://howto.eurec4a.eu/running_locally.html) a bit of a description, how the book is meant to work. If you follow that description, you'll see if the article appears correctly on the generated pate.

Could you converte the notebook to `MySt` format and remove the generated plots and output from the repository? They'll get regenerated if the code works out.

Further, references should go into [references.bib](https://github.com/eurec4a/how_to_eurec4a/blob/master/how_to_eurec4a/references.bib) and the article would have to be included somewhere in [_toc.yml](https://github.com/eurec4a/how_to_eurec4a/blob/master/how_to_eurec4a/_toc.yml) to show up in the final page.

lvol08 commented:
Hey @d70-t , thanks for the comments and sorry for not noticing in advance. 

I changed the notebook to MySt format and added the references to the references.bib and included the notebook in the _toc.yml file. However, I am not sure if the notebook actually works with the given requirements for the python environment but that's unfortunately the case for all notebooks in the catalogue that use matplotlib for me. I can install matplotlib again in the environment but shouldn't it work just with installing the requirements?

d70-t commented on how_to_eurec4a/specmacs_cloud_geometry.md:
````diff
@@ -0,0 +1,267 @@
+```
+
+```{code-cell} ipython3
+#create norm for scatter plot
+cmap = 'OrRd'
+vmin = np.min([np.min(height), np.min(height_above_geoid)])
+vmax = np.max([np.max(height), np.max(height_above_geoid)])
+norm = matplotlib.colors.Normalize(vmin=vmin, vmax=vmax)
+sm = matplotlib.cm.ScalarMappable(norm=norm, cmap=cmap)
````
I'd suggest to simplify the `min` and `max` calculation.
I'd also probably use `vmin` and `vmax` instead of `norm`. While this is less flexible, it's probably more commonly known. In order to get the colorbars right, the result of one of the `scatter` calls can be assigned to `sm`.
```suggestion
cmap = 'OrRd'
vmin = np.min([height, height_above_geoid])
vmax = np.max([height, height_above_geoid])
```

d70-t commented on how_to_eurec4a/specmacs_cloud_geometry.md:
````diff
@@ -0,0 +1,267 @@
+
+```{code-cell} ipython3
+fig, axs = plt.subplots(nrows=2, ncols=1, figsize=(20, 8), sharex=True)
+axs[0].scatter(lon, lat, c=height, marker='o', s=2, edgecolor=None, linewidth=0, norm=norm, cmap=cmap)
+axs[0].set_title('cloud top height above WGS84 in m')
+axs[0].set_ylabel('latitude [°N]')
+divider = make_axes_locatable(axs[0])
+ax = divider.append_axes('right', size='40%', pad=0.05)
+ax.hist(height, bins=np.linspace(vmin, vmax, 100), density=True)
+ax.set_title('probability density')
+ax.yaxis.tick_right()
+ax.axvline(np.mean(height), linestyle='dashed', color='k', label='mean {:.2f} m'.format(np.mean(height)))
+ax.legend()
+axs[1].scatter(lon, lat, c=height_above_geoid, marker='o', s=2, edgecolor=None, linewidth=0, norm=norm, cmap=cmap)
+axs[1].set_title('cloud top height above sea surface in m')
+axs[1].set_ylabel('latitude [°N]')
+axs[1].set_xlabel('longitude [°E]')
+divider = make_axes_locatable(axs[1])
+ax = divider.append_axes('right', size='40%', pad=0.05)
+ax.hist(height_above_geoid, bins=np.linspace(vmin, vmax, 100), density=True)
+ax.set_title('probability density')
+ax.yaxis.tick_right()
+ax.axvline(np.mean(height_above_geoid), linestyle='dashed', color='k', label='mean {:.2f} m'.format(np.mean(height_above_geoid)))
+ax.legend()
+plt.colorbar(sm, ax=axs)
````
Much of this code is repeated for both rows of the plot. I suggest to put it inside a loop, to clarify how this code is executed twice.

For the `hist`, I also recommend to pass in the number of bins and the (equidistant) range explicitly. The argument is similar to `isel` above: this way, the `hist` function knows that the bins are equidistant and can use a more optimal code path.
```suggestion
for ax, var, ref in zip(axs, [height, height_above_geoid], ["WGS84", "sea surface"]):
    sm = ax.scatter(lon, lat, c=var, marker='o', s=2, edgecolor=None, linewidth=0, vmin=vmin, vmax=vmax, cmap=cmap)
    ax.set_title(f'cloud top height above {ref} in m')
    ax.set_ylabel('latitude [°N]')
    divider = make_axes_locatable(ax)
    ax2 = divider.append_axes('right', size='40%', pad=0.05)
    ax2.hist(height, bins=100, range=(vmin, vmax), density=True)
    ax2.set_title('probability density')
    ax2.yaxis.tick_right()
    ax2.axvline(np.mean(var), linestyle='dashed', color='k', label='mean {:.2f} m'.format(np.mean(var)))
    ax2.legend()

axs[1].set_xlabel('longitude [°E]')
plt.colorbar(sm, ax=axs)
```

d70-t commented on how_to_eurec4a/specmacs_cloud_geometry.md:
````diff
@@ -0,0 +1,267 @@
+ax.axvline(np.mean(height_above_geoid), linestyle='dashed', color='k', label='mean {:.2f} m'.format(np.mean(height_above_geoid)))
+ax.legend()
+plt.colorbar(sm, ax=axs)
+```
````
I'd recommend to add a `None` statement (or a semicolon `;` after the last command) to suppres the printout of the colorbar object.
````suggestion
None
```
````

d70-t commented on how_to_eurec4a/specmacs_cloud_geometry.md:
````diff
@@ -0,0 +1,267 @@
+import matplotlib.ticker as ticker
+```
+
+```{code-cell} ipython3
+#settings for plots 
````
Again: text is better placed in the markdown section.
```suggestion
settings for plots
```{code-cell} ipython3
```

d70-t commented on how_to_eurec4a/specmacs_cloud_geometry.md:
````diff
@@ -0,0 +1,267 @@
+
+```{code-cell} ipython3
+#settings for plots 
+s = 0.1 #markersize
````
I suggest to use a telling variable name instead of a hard to remember variable name and a comment.
```suggestion
markersize = 0.1
```

d70-t commented on how_to_eurec4a/specmacs_cloud_geometry.md:
```diff
@@ -0,0 +1,267 @@
+cmap = 'viridis' #colormap
+vmin = np.min(height)
+vmax = np.max(height)
+norm = matplotlib.colors.Normalize(vmin=vmin, vmax=vmax)
+sm = matplotlib.cm.ScalarMappable(norm=norm, cmap=cmap)
```
(as before)
```suggestion
```

d70-t commented on how_to_eurec4a/specmacs_cloud_geometry.md:
````diff
@@ -0,0 +1,267 @@
+
+```{code-cell} ipython3
+fig, axs = plt.subplots(2, 2, figsize=(14,12), subplot_kw=dict(projection='3d'))
+
````
I was wondering if there could be a good way to reduce repetitions in in the code of the following plot, but I didn't come up with something much better. I'm however wondering if it might be more readable to do the actuall `scatter` in a loop, just to emphasize that it's really always the same data:

(this change would also account for `vmin`, `vmax` and `markersize`).

```suggestion

for ax in axs.ravel():
    sm = ax.scatter(lon, lat, height, c=height, marker='o', vmin=vmin, vmax=vmax, cmap=cmap, s=markersize, linewidths=None, edgecolors=None)

```

d70-t commented on how_to_eurec4a/specmacs_cloud_geometry.md:
```diff
@@ -0,0 +1,267 @@
+fig, axs = plt.subplots(2, 2, figsize=(14,12), subplot_kw=dict(projection='3d'))
+
+ax = axs[0,0]
+ax.scatter(lon, lat, height, c=height, marker='o', norm=norm, cmap=cmap, s=s, linewidths=None, edgecolors=None)
```
```suggestion
```

d70-t commented on how_to_eurec4a/specmacs_cloud_geometry.md:
```diff
@@ -0,0 +1,267 @@
+ax.set_title('3D stereo plot', y=1.02)
+
+ax = axs[0,1]
+ax.scatter(lon, lat, height, c=height, marker='o', norm=norm, cmap=cmap, s=s, linewidths=None, edgecolors=None)
```
```suggestion
```

d70-t commented on how_to_eurec4a/specmacs_cloud_geometry.md:
```diff
@@ -0,0 +1,267 @@
+ax.xaxis.set_major_locator(ticker.NullLocator())
+
+ax = axs[1,0]
+ax.scatter(lon, lat, height, c=height, marker='o', norm=norm, cmap=cmap, s=s, linewidths=None, edgecolors=None)
```
```suggestion
```

d70-t commented on how_to_eurec4a/specmacs_cloud_geometry.md:
```diff
@@ -0,0 +1,267 @@
+ax.zaxis.set_major_locator(ticker.NullLocator())
+
+ax = axs[1,1]
+ax.scatter(lon, lat, height, c=height, marker='o', norm=norm, cmap=cmap, s=s, linewidths=None, edgecolors=None)
```
```suggestion
```

d70-t commented on how_to_eurec4a/specmacs_cloud_geometry.md:
```diff
@@ -0,0 +1,267 @@
+
+plt.subplots_adjust(bottom=0.1, right=0.9)
+cax = plt.axes([0.95, 0.15, 0.05, 0.7])
+fig.colorbar(sm, cax=cax, label='height above WGS84 [m]', fraction=0.02, shrink=2.0, aspect=40)
```
```suggestion
fig.colorbar(sm, cax=cax, label='height above WGS84 [m]', fraction=0.02, shrink=2.0, aspect=40)
None
```

d70-t commented on how_to_eurec4a/specmacs_cloud_geometry.md:
````diff
@@ -0,0 +1,267 @@
+```{code-cell} ipython3
+class DefaultTileWTS(GoogleWTS):
+    def get_image(self, tile):
+        if six.PY3:
+            from urllib.request import urlopen, Request, HTTPError, URLError
+        else:
+            from urllib2 import urlopen, Request, HTTPError, URLError
+
+        url = self._image_url(tile)
+        try:
+            request = Request(url)
+            fh = urlopen(request)
+            im_data = six.BytesIO(fh.read())
+            fh.close()
+            img = Image.open(im_data)
+
+        except (HTTPError, URLError) as err:
+            img = self._default_tile(tile)
+
+        img = img.convert(self.desired_tile_form)
+        return img, self.tileextent(tile), 'lower'
+
+class specMACSPTiles(DefaultTileWTS):
+    def __init__(self, flight_segment_id):
+        super(specMACSPTiles, self).__init__(desired_tile_form="RGBA")
+        self.flight_segment_id = flight_segment_id
+        self.empty_tile = Image.fromarray(np.full((256, 256, 4), (255, 255, 255, 0), dtype=np.uint8))
+
+    def _default_tile(self, tile):
+        return self.empty_tile
+
+    def _image_url(self, tile):
+        x, y, z = tile
+        return f"https://macsserver.physik.uni-muenchen.de/campaigns/EUREC4A/maps/tiles/{self.flight_segment_id}/{z}/{x}/{y}.png"
+```
+
+```{code-cell} ipython3
+#get specMACS tile
+specmacs = specMACSPTiles(segment['segment_id'])
+```
````
I'd try to keep the required code for the tiles rather minimal. I'd consider the tiles as a very nice background, but in principle unrelated to the purpose of this article, namely the stereo retrieval results. Thus I'd
* drop Python 2 support and the dependency on `six`
* merge the two classes into one
* hide the code by default
* move `cartopy.crs` import further down, as it's unrelated to this (now hidden) code block, but important for the following code

This change will also account for `segment_id` from the beginning.
````suggestion
We will start to project the points found from the stereographic reconstruction to pre-rendered specMACS [tile maps](https://en.wikipedia.org/wiki/Tiled_web_map) which can be found [here](https://macsserver.physik.uni-muenchen.de/campaigns/EUREC4A/maps/).
To make them available to `cartopy`, we need a little helper class `specMACSPTiles`, which mainly takes care of making unavailable tiles transparent (these are outside the viewing range of the instrument).

```{code-cell} ipython3
:tags: [hide-cell]

from cartopy.io.img_tiles import GoogleWTS
from PIL import Image

class specMACSPTiles(GoogleWTS):
    def __init__(self, flight_segment_id):
        super().__init__(desired_tile_form="RGBA")
        self.flight_segment_id = flight_segment_id
        self.empty_tile = Image.fromarray(np.full((256, 256, 4), (255, 255, 255, 0), dtype=np.uint8))

    def get_image(self, tile):
        from io import BytesIO
        from urllib.request import urlopen, Request, HTTPError, URLError

        url = self._image_url(tile)
        try:
            request = Request(url)
            fh = urlopen(request)
            im_data = BytesIO(fh.read())
            fh.close()
            img = Image.open(im_data)

        except (HTTPError, URLError) as err:
            img = self.empty_tile

        img = img.convert(self.desired_tile_form)
        return img, self.tileextent(tile), 'lower'

    def _image_url(self, tile):
        x, y, z = tile
        return f"https://macsserver.physik.uni-muenchen.de/campaigns/EUREC4A/maps/tiles/{self.flight_segment_id}/{z}/{x}/{y}.png"
```

We can now get the tile map for our selected flight segment:

```{code-cell} ipython3
specmacs = specMACSPTiles(segment_id)
```
````

d70-t commented on how_to_eurec4a/specmacs_cloud_geometry.md:
````diff
@@ -0,0 +1,267 @@
+specmacs = specMACSPTiles(segment['segment_id'])
+```
+
+```{code-cell} ipython3
+#get vmin and vmax using quantiles such that outliers will not influence the colorbar too much
+q = 0.98
+vmin = np.quantile(height, (1-q)/2)
+vmax = np.quantile(height, q+(1-q)/2)
+
+cmap = 'OrRd'
+norm = matplotlib.colors.Normalize(vmin=vmin, vmax=vmax)
+sm = matplotlib.cm.ScalarMappable(norm=norm, cmap=cmap)
+```
````
I'd suggest to move the comment into markdown, simplify the quantile computation and do the `vmin`, `vmax` trick again.
````suggestion
Get `vmin` and `vmax` using quantiles such that outliers will not influence the colorbar too much:
```{code-cell} ipython3
q = 0.01
vmin, vmax = np.quantile(height, [q, 1-q])
cmap = 'OrRd'
```
````

d70-t commented on how_to_eurec4a/specmacs_cloud_geometry.md:
````diff
@@ -0,0 +1,267 @@
+```
+
+```{code-cell} ipython3
+fig = plt.figure(figsize=(16, 8), constrained_layout=True)
+ax = fig.add_subplot(1, 2, 1, projection=ccrs.PlateCarree())
+ax.set_extent([np.floor(np.min(lon)), np.ceil(np.max(lon)), np.floor(np.min(lat)), np.ceil(np.max(lat))], crs=ccrs.PlateCarree())
+ax.add_image(specmacs, 8) # the number is an integer and represents the detail level. It is logarithmic, small values are coarse, large values are fine.
+ax.set_axis_off()
+ax = fig.add_subplot(1, 2, 2, projection=ccrs.PlateCarree())
+ax.set_extent([np.floor(np.min(lon)), np.ceil(np.max(lon)), np.floor(np.min(lat)), np.ceil(np.max(lat))], crs=ccrs.PlateCarree())
+ax.add_image(specmacs, 8) # the number is an integer and represents the detail level. It is logarithmic, small values are coarse, large values are fine.
+ax.set_axis_off()
+im = ax.scatter(lon, lat, c=height, marker="o", s=0.4, edgecolor=None, linewidth=0, norm=norm, cmap=cmap)
````
Here, I'd again suggest to place the plotting inside a loop, just to have less repeated code.

Also, this is the place where `cartopy.crs` is needed, I'd place it here such that it's not hidden.

````suggestion
import cartopy.crs as ccrs

fig, axes = plt.subplots(1, 2, figsize=(16, 8), constrained_layout=True, subplot_kw={"projection": ccrs.PlateCarree()})
for ax in axes:
    ax.set_extent([np.floor(np.min(lon)), np.ceil(np.max(lon)), np.floor(np.min(lat)), np.ceil(np.max(lat))], crs=ccrs.PlateCarree())
    ax.add_image(specmacs, 8)  # the number is an integer and represents the detail level. It is logarithmic, small values are coarse, large values are fine.
    ax.set_axis_off()
im = ax.scatter(lon, lat, c=height, marker="o", s=0.4, edgecolor=None, linewidth=0, vmin=vmin, vmax=vmax, cmap=cmap)
````

d70-t commented on how_to_eurec4a/specmacs_cloud_geometry.md:
```diff
@@ -0,0 +1,267 @@
+ax.add_image(specmacs, 8) # the number is an integer and represents the detail level. It is logarithmic, small values are coarse, large values are fine.
+ax.set_axis_off()
+im = ax.scatter(lon, lat, c=height, marker="o", s=0.4, edgecolor=None, linewidth=0, norm=norm, cmap=cmap)
+fig.colorbar(im, label='height above WGS84 [m]', extend='both', shrink=0.4)
```
```suggestion
fig.colorbar(im, label='height above WGS84 [m]', extend='both', shrink=0.4)
None
```

d70-t reviewed (commented):
Thanks again for this very nice article!
The amount of comments might seem large, but I'd really see almost all of them as very minor stylistic updates. The idea behind these changes is to try to reduce the amount of code unrelated to the main purpose of this article (the cloud geometry dataset), such that it becomes (even) easier to read and to keep the focus on the important things.

I hope the suggested changes are consistent, but I don't know a way of testing the suggestions upfront.

d70-t commented:
One more minor thing:
Before merging, we should probably do a rebase, removing the previously committed figures, to keep the repo small.

lvol08 commented:
Hi @d70-t ,
thanks a lot for all the suggestions to make the code more simple and better readable! :) I think I included all the suggestions in the commits. About the rebase: Do I have to do it in my local fork, so something like "git rebase master" and place a new merge request afterwards? 
Thanks again!

d70-t commented:
> About the rebase: Do I have to do it in my local fork, so something like "git rebase master" and place a new merge request afterwards?

You can do it in your local fork, or I can do it for you. In any case, the result would be "force pushed" to the current branch ([`lvol08:specmacs_cloud_geometry`](https://github.com/lvol08/how_to_eurec4a/tree/specmacs_cloud_geometry)) and thus the current pull request will get updated. In particular, we wouldn't want to have a new pull request, because that way we'd loose all the discussion up to now.

It would probably be an interactive rebase, e.g.:
```
git rebase -i master
```
(or `origin/master` or `upstream/master` or whatever the way you'd locally reference the current `master` of the main online repository)

In this [interactive rebase](https://git-scm.com/docs/git-rebase#_interactive_mode) you could *squash* or *fixup* all the commits into a single one. Afterwards, the branch would be "diverged" from the copy which is currently referenced in the pull request (because it doesn't contain the individual commits anymore) and thus to upload (and thereby replace the current commits), you'd have to use `git push -f` (aka *force*).

d70-t commented:
The updates are working fine 🎉. The build is still failing due to #102, but that's fine for this pull request.
After squashing the commits (i.e. using a rebase), it's good to go 😄.
@lvol08 you decide who'll do the squash.

lvol08 commented:
Thanks @d70-t ! I fixed one more typo, did the rebase and the force push, hope that it works all out :)

d70-t commented:
I just did another rebase on the (again) updated master, thus the ongoing fixes from the other issues are now incorporated and the pipeline is finally green :-)

d70-t commented:
Haha, I screwed it up... Will fix it it a moment 🤦
